File: rgctl-tests/ecommerce-go/internal/service/product.go
package service

import (
	"github.com/google/uuid"
	"github.com/sshaaf/ecommerce-go/internal/models"
	"github.com/sshaaf/ecommerce-go/internal/pkg/timeutil"
	"github.com/sshaaf/ecommerce-go/internal/repository"
)

type ProductService struct {
	products *repository.ProductRepository
}

func NewProductService(products *repository.ProductRepository) *ProductService {
	return &ProductService{products: products}
}

type CreateProductRequest struct {
	CategoryID  string `json:"category_id" binding:"required"`
	Name        string `json:"name" binding:"required"`
	Slug        string `json:"slug" binding:"required"`
	Description string `json:"description"`
	PriceCents  int64  `json:"price_cents" binding:"required,min=0"`
	Stock       int64  `json:"stock" binding:"min=0"`
}

type ProductResponse struct {
	ID          string `json:"id"`
	CategoryID  string `json:"category_id"`
	Name        string `json:"name"`
	Slug        string `json:"slug"`
	Description string `json:"description"`
	PriceCents  int64  `json:"price_cents"`
	Stock       int64  `json:"stock"`
}

func (s *ProductService) List() ([]ProductResponse, error) {
	products, err := s.products.List()
	if err != nil {
		return nil, err
	}
	out := make([]ProductResponse, len(products))
	for i, p := range products {
		out[i] = toProductResponse(p)
	}
	return out, nil
}

func (s *ProductService) Get(id string) (*ProductResponse, error) {
	product, err := s.products.FindByID(id)
	if err != nil {
		return nil, MapRepoError(err)
	}
	resp := toProductResponse(*product)
	return &resp, nil
}

func (s *ProductService) Create(req CreateProductRequest) (*ProductResponse, error) {
	product := &models.Product{
		ID:          uuid.NewString(),
		CategoryID:  req.CategoryID,
		Name:        req.Name,
		Slug:        req.Slug,
		Description: req.Description,
		PriceCents:  req.PriceCents,
		Stock:       req.Stock,
		CreatedAt:   timeutil.NowISO(),
	}
	if err := s.products.Create(product); err != nil {
		return nil, err
	}
	resp := toProductResponse(*product)
	return &resp, nil
}

func toProductResponse(p models.Product) ProductResponse {
	return ProductResponse{
		ID:          p.ID,
		CategoryID:  p.CategoryID,
		Name:        p.Name,
		Slug:        p.Slug,
		Description: p.Description,
		PriceCents:  p.PriceCents,
		Stock:       p.Stock,
	}
}

File: rgctl-tests/ecommerce-python/app/services/product.py
import uuid

from sqlalchemy.orm import Session

from app.exceptions import ConflictError, NotFoundError
from app.models.product import Product
from app.repositories.category import CategoryRepository
from app.repositories.product import ProductRepository
from app.schemas.product import CreateProductRequest, ProductResponse


class ProductService:
    def __init__(self, db: Session) -> None:
        self.products = ProductRepository(db)
        self.categories = CategoryRepository(db)

    def list_all(self) -> list[ProductResponse]:
        return [ProductResponse.model_validate(p) for p in self.products.list_all()]

    def get(self, product_id: str) -> ProductResponse:
        product = self.products.get(product_id)
        if not product:
            raise NotFoundError("Product not found")
        return ProductResponse.model_validate(product)

    def create(self, req: CreateProductRequest) -> ProductResponse:
        if not self.categories.get(req.category_id):
            raise NotFoundError("Category not found")
        if self.products.find_by_slug(req.slug):
            raise ConflictError("slug already exists")
        product = Product(
            id=str(uuid.uuid4()),
            category_id=req.category_id,
            name=req.name,
            slug=req.slug,
            description=req.description,
            price_cents=req.price_cents,
            stock=req.stock,
        )
        self.products.add(product)
        return ProductResponse.model_validate(product)
